mirgecom/artificial_viscosity.py

```diff
@@ -0,0 +1,346 @@
+            & s_0-\kappa \le s_e \le s_0+\kappa \\
+            1, & s_e > s_0+\kappa
+        \end{cases}
+
+where:
+
+- $\varepsilon_e$ is the element viscosity
+- $s_e = \log_{10}{S_e} \sim 1/p^4$ is the smoothness indicator
+- $s_0$ is a reference smoothness value
+- $\kappa$ controls the width of the transition between 0 to 1
+
+Smoothness Indicator Evaluation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. autofunction:: smoothness_indicator
+
+AV RHS Evaluation
+^^^^^^^^^^^^^^^^^
+
+.. autofunction:: av_operator
+"""
+
+__copyright__ = """
+Copyright (C) 2020 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+import loopy as lp
+from modepy import vandermonde
+from pytools.obj_array import obj_array_vectorize
+from meshmode.dof_array import thaw, DOFArray
+from meshmode.mesh import BTAG_ALL, BTAG_NONE  # noqa
+from grudge.eager import interior_trace_pair, cross_rank_trace_pairs
+from grudge.symbolic.primitives import TracePair
+from mirgecom.fluid import (
+    split_conserved,
+    join_conserved_vectors
+)
+
+
+def _facial_flux_q(discr, q_tpair):
+    """Compute facial flux for each scalar component of Q."""
+    q_int = q_tpair.int
+    actx = q_int[0].array_context
+
+    flux_dis = q_tpair.avg
+    if isinstance(flux_dis, np.ndarray):
+        flux_dis = flux_dis.reshape(-1, 1)
+
+    normal = thaw(actx, discr.normal(q_tpair.dd))
+
+    # This uses a central scalar flux along nhat:
+    # flux = 1/2 * (Q- + Q+) * nhat
+    flux_out = flux_dis * normal
+
+    return discr.project(q_tpair.dd, "all_faces", flux_out)
+
+
+def _facial_flux_r(discr, r_tpair):
+    """Compute facial flux for vector compontnent of grad(Q)."""
+    r_int = r_tpair.int
+    actx = r_int[0][0].array_context
+
+    normal = thaw(actx, discr.normal(r_tpair.dd))
+
+    # This uses a central vector flux along nhat:
+    # flux = 1/2 * (grad(Q)- + grad(Q)+) .dot. nhat
+    flux_out = r_tpair.avg @ normal
```

Comment:
From my perspective, no.  The code expressions closely match the expressions in the BR1 paper and in Hesthaven.   Hitting it with further python magic (imo) obfuscates this correspondence, and renders the function harder to read by domain users.

Reply:
This is kind of what I had mind, FWIW: https://github.com/majosm/mirgecom/commit/80ec49f38d6859d9ebc55d73865848caef8b85de

Reply:
Yeah, so I do get that your way is more pythonic.  But to me, the previous construction without the fancy stuff was more straightforward and more easily understood.

Reply:
Re: Why the shift to a fluid-specific operator for AV.

The previous AV operator was arranged for doing an arbitrary number of independent scalar equations dropped into a general "Q".  But the fluid equations are coupled, i.e. the fluid calculations (including boundary treatments, fluxes, EOS, etc) all require ALL components of Q simultaneously for calculations.  

It is difficult to marry the general, uncoupled Q treatment of the previous AV operator/driver with the fluid-specific coupled treatment.  Some code gymnastics (both in the operator, and in the fluid boundary treatments) were being applied to handle/assemble the required flux data at element boundaries, and it also required user-knowledge of Q structure (Q[0]=mass, Q[1]=energy), etc.).

After discussion with @w-hagen and @inducer, we (at this time) are leaning towards having a separate fluid-specific operator and a different operator for AV on uncoupled scalar components.  The resulting fluid-specific AV operator is consistent with the formulation, and with the documentation (mostly), because both specifically target the Euler eqns.

Reply:
This is complete garbage of course. @majosm fix above allowed re-generalization of the AV and will show up when w-hagen/mirgecom#16 lands.